### ISSUE TITLE
Allow self-contained knitr_print

### ISSUE THREAD

**Issue by elbamos** [ramnathv/htmlwidgets#146]
I'm trying to make it possible to embed htmlwidgets in a larger project using knitr.  It seems that the knitr_print function for HTML widgets sets standalone=false, which makes it a lot harder to embed. Is it possible to change this so it pulls standalone from an option or from the chunk options? Or is FALSE hardcoded because that just won't work?

**Comment by ramnathv** [comment]
If you set `self_contained: true` in the YAML header of your `rmarkdown` document, it will automatically embed a standalone version of the widget in the output html.

**Comment by elbamos** [comment]
Right. But, I want to use knitr rather than rmarkdown.

**Comment by ramnathv** [comment]
Can you provide a simple reproducible example to illustrate the issue at hand? If you are NOT using `rmarkdown`, what document are you trying to knit?

**Comment by elbamos** [comment]
I can try to later today. I'm writing an R plugin for Apache Zeppelin, which is a notebook project similar to Jupyter but oriented toward spark and other large data applications.  (See my rinterpreter repo for slightly aged code.) I have to return an HTML fragment for insertion into the page.  This is easier and faster to do with knitr than rmarkdown because knitr has lower overhead and I believe generally doesn't need to read and write temp files to disc. (I recognize that some of that benefit may be lost with htmlwidgets because it may need to call pandoc.) So I run knitr, then parse the HTML to extract the tags inside of <body>, adding back any <script> tags inside the head, and insert that into the notebook cell.  I have display of googleVis and rCharts working. I'd like to get htmlwidgets running too.

(Figuring out how to integrate with shiny/ggvis, is a project for the future.)

Thanks for your attention!

**Comment by ramnathv** [comment]
Ah. Interesting. Jupyter notebooks have a different dependency management mechanism that involves `requirejs`. I have been thinking about how to get htmlwidgets to work natively in notebooks, but the `requirejs` stuff makes it a little complicated.

One easy solution is to use `iframes`. Will that work for you? I have a pending PR that makes it easy to render a widget as a self-contained `iframe`, and we can test if it works for your use-case.

**Comment by elbamos** [comment]
Wouldn't an iframe require linking rather than embedding html? In general, I think iframes are kinda a last resort solution, and I'd prefer not to move all use of knitr to iframes for this one purpose. 

Would the as.tags() function do what I need, even if not in knitr? If it does provide the right output, one possibility is to try to overload knit_print.htmlwidget. And if that's right, then this seems like it would be pretty simple as a modification to knit_print.

Good to hear that htmlwidgets don't work in Jupyter--so if we get it to work, Zeppelin will be first.

**Comment by ramnathv** [comment]
Well you can embed iframes inline using the `srcref` attribute. The downside is that it doesn't work in all browsers. Based on what you indicated, it seems to me that what we need is a modified version of `htmltools::save_html`, which is the underlying force behind `htmlwidgets::saveWidget`.

```r
save_html = function (html, file, libdir = "lib") {
    dir <- dirname(file)
    oldwd <- setwd(dir)
    on.exit(setwd(oldwd), add = TRUE)
    rendered <- renderTags(html)
    deps <- lapply(rendered$dependencies, function(dep) {
        dep <- copyDependencyToDir(dep, libdir, FALSE)
        dep <- makeDependencyRelative(dep, dir, FALSE)
        dep
    })
    html <- c("<!DOCTYPE html>", "<html>", "<head>", "<meta charset=\"utf-8\"/>", 
        renderDependencies(deps, c("href", "file")), rendered$head, 
        "</head>", "<body>", rendered$html, "</body>", "</html>")
    writeLines(html, file, useBytes = TRUE)
}
```

Based on your description, seems like you need a `htmltools:::save_html_fragment`, that would modify the above function and write 

```
html <- c(renderDependencies(deps, c("href", "file")), rendered$html, "</body>")
```

You would then need a modified version of `htmlwidgets::saveWidget` that calls `htmltools::save_html_fragment` instead of `htmltools::save_html`.

If you think this would work, I would be happy to add this functionality to `htmlwidgets`.

**Comment by ramnathv** [comment]
This is related to [https://github.com/ramnathv/htmlwidgets/issues/144].

**Comment by elbamos** [comment]
Looks perfect to me! How can I help?

Thanks!

**Comment by elbamos** [comment]
Actually just to make sure — would it use that automatically inside of a knitr chunk?  And is it self-contained?


From: AMOS ELBERG <amos.elberg@me.com>
Reply: AMOS ELBERG <amos.elberg@me.com>
Date: August 10, 2015 at 5:50:45 PM
To: ramnathv/htmlwidgets <reply@reply.github.com>
Cc: ramnathv/htmlwidgets <htmlwidgets@noreply.github.com>
Subject:  Re: [htmlwidgets] Allow self-contained knitr_print (#146)  

Looks perfect to me! How can I help?

Thanks!

On Aug 10, 2015, at 5:29 PM, Ramnath Vaidyanathan <notifications@github.com> wrote:

Well you can embed iframes inline using the srcref attribute. The downside is that it doesn't work in all browsers. Based on what you indicated, it seems to me that what we need is a modified version of htmltools::save_html, which is the underlying force behind htmlwidgets::saveWidget.

save_html = function (html, file, libdir = "lib") {
    dir <- dirname(file)
    oldwd <- setwd(dir)
    on.exit(setwd(oldwd), add = TRUE)
    rendered <- renderTags(html)
    deps <- lapply(rendered$dependencies, function(dep) {
        dep <- copyDependencyToDir(dep, libdir, FALSE)
        dep <- makeDependencyRelative(dep, dir, FALSE)
        dep
    })
    html <- c("<!DOCTYPE html>", "<html>", "<head>", "<meta charset=\"utf-8\"/>",  
        renderDependencies(deps, c("href", "file")), rendered$head,  
        "</head>", "<body>", rendered$html, "</body>", "</html>")
    writeLines(html, file, useBytes = TRUE)
}
Based on your description, seems like you need a htmltools:::save_html_fragment, that would modify the above function and write

html <- c(renderDependencies(deps, c("href", "file")), rendered$html, "</body>")

You would then need a modified version of htmlwidgets::saveWidget that calls htmltools::save_html_fragment instead of htmltools::save_html.

If you think this would work, I would be happy to add this functionality to htmlwidgets.

**Comment by ramnathv** [comment]
First goal will be to come up with a function, `renderInZeppelin(w)` that would render a widget object `w` in an Apache Zeppelin notebook. Once we have that covered, we can think of how to make it work without having to make an extra call.

**Comment by elbamos** [comment]
Sounds right; would you mind if I do that part of the code so its apache licensed?

**Comment by elbamos** [comment]
One of the things that has to happen for this to work, is converting the dependencies by either linking to an external website, or (more likely) encoding with Base64 and embedding in the HTML. Do you know of any existing R implementation of that, or have any suggestions?

It seems like something that might be a PR for htmltools if it isn't already lying around in a package somewhere.

**Comment by ramnathv** [comment]
This is exactly what `rmarkdown` does using `pandoc`. It takes each of the js/css assets and converts them into a base64 encoded dataURI. The nice thing with `pandoc` is that it does this recursively, as a result of which the dataURIs are self-contained.

I have an R implementation that does all of this EXCEPT for recursive handling of assets. See [here](https://github.com/rstudio/htmltools/pull/9).

If you are interested in tackling the recursive handling of assets, let me know. It would be a very valuable addition.

**Comment by elbamos** [comment]
Yes I can take a look at this, you're right. Is recursiveness required for Js though or just for css? I would think that a Js-based visualization will run fine even if the Css is completely missing but that may be my misunderstanding of front-end stuff.

**Comment by jjallaire** [comment]
It's recursiveness for everything (CSS can reference fonts and other
stylesheets).

**Comment by elbamos** [comment]
What happens to htmlwidgets if the Css isn't there? Do they fail to run or just look poorly formatted?

**Comment by ramnathv** [comment]
Well, they would look very poorly formatted, and in some cases not display everything. So I would NEVER use a widget without the underlying css.

**Comment by elbamos** [comment]
Thanks - super helpful. I'll give a try to writing something tonight. If you have a good test case that would help.

**Comment by elbamos** [comment]
Ramnath, FYI I have not given up on doing this. I am, however, now understanding why it's tricky! Would you happen to have a few well-understood use cases I could use for testing? I need ones with and without recursive scripts... I suspect what's happening in my current code is that it's working when the scripts are flat but not when they're recursive.

**Comment by ramnathv** [comment]
@elbamos I understand. The recursive assets are tricky, since you first need to locate these resources and then inline them as data URIs. I would think that the `bootstrap` framework would be a good test case, since the CSS links to external font files.

**Comment by jjallaire** [comment]
Note that all of this would be unnecessary if you were willing to use
pandoc (as it does all of the recursive embedding you are looking for).
Since rmarkdown is basically knitr + pandoc (and pandoc just renders
markdown and does recursive base64 embedding) + htmlwidgets it seems like
it does everything you need and trying to painstakingly reproduce it might
be misdirected effort.

Of course I may have missed something important here, just putting in my 2
cents based on observing from afar.

**Comment by elbamos** [comment]
Yeah I've been thinking about that. Effectively it would mean using rmarkdown instead of knitr. Besides introducing a dependency, when I tried it that way, it was very, very slow... But I've been thinking about it. I put out a stack overflow today asking if anyone knows of a Java library that does it.

**Comment by jjallaire** [comment]
Yes, pandoc can definitely be slow and that is a clearly a downside if you
want the rendering to occur in sub 1 or 2 seconds.

**Comment by elbamos** [comment]
Well I'm trying to serve html live. Pandoc is a lot of overhead. The html has to get written to disc, start the Pandoc process, more disc, then read it back for serving. If I can do this in R then theres no disc or separate process at all.
